File: solutions/python3/0126.py
class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        def dfs(word: str, path: List[str]) -> None:
            if word == endWord:
                ans.append(path)
                return
            if word not in dict:
                return

            for child in dict[word]:
                dfs(child, path + [child])

        ans = []
        set_ = set(wordList)

        if endWord not in set_:
            return ans

        set1 = set([beginWord])
        dict = collections.defaultdict(list)
        isFound = False

        while set1 and not isFound:
            for word in set1:
                set_.discard(word)
            tempSet = set()
            for parent in set1:
                for i in range(len(parent)):
                    for j in string.ascii_lowercase:
                        newWord = parent[:i] + j + parent[i + 1:]
                        if newWord == endWord:
                            dict[parent].append(newWord)
                            isFound = True
                        elif newWord in set_ and not isFound:
                            tempSet.add(newWord)
                            dict[parent].append(newWord)
            set1 = tempSet

        if isFound:
            dfs(beginWord, [beginWord])

        return ans

File: solutions/cpp/0126.cpp
class Solution {
 public:
  vector<vector<string>> findLadders(string beginWord, string endWord,
                                     vector<string>& wordList) {
    vector<vector<string>> ans;
    unordered_set<string> set{begin(wordList), end(wordList)};

    if (!set.count(endWord)) return ans;

    unordered_set<string> set1{beginWord};
    unordered_map<string, vector<string>> map;
    bool isFound = false;

    while (!set1.empty() && !isFound) {
      for (const string& word : set1) set.erase(word);
      unordered_set<string> tempSet;
      for (const string& parent : set1) {
        string word = parent;
        for (int i = 0; i < word.length(); ++i) {
          char c = word[i];
          for (char j = 'a'; j <= 'z'; ++j) {
            word[i] = j;
            if (word == endWord) {
              map[parent].push_back(word);
              isFound = true;
            } else if (set.count(word) && !isFound) {
              tempSet.insert(word);
              map[parent].push_back(word);
            }
          }
          word[i] = c;
        }
      }
      swap(set1, tempSet);
    }

    if (isFound) {
      vector<string> path{beginWord};
      dfs(map, beginWord, endWord, path, ans);
    }

    return ans;
  }

 private:
  void dfs(const unordered_map<string, vector<string>>& map, const string& word,
           const string& endWord, vector<string>& path,
           vector<vector<string>>& ans) {
    if (word == endWord) {
      ans.push_back(path);
      return;
    }
    if (!map.count(word)) return;

    for (const string& child : map.at(word)) {
      path.push_back(child);
      dfs(map, child, endWord, path, ans);
      path.pop_back();
    }
  }
};